<a href="https://colab.research.google.com/github/Toulouse08/IS-170-Machine-Learning/blob/Association-Rules/Lab5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Association Rules**

# Using 'lab5.CSV'

David Bautista 
Edwin Burrell 
Corry Fuentes

In [ ]:
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules


In [ ]:
df=pd.read_csv('lab5.csv')
df.head()

Sanitizer, Wetwipes, Bread
0   Sanitizer, Wetwipes, Juice
1    Slippers, Wetwipes, Juice
2  Sanitizer, Bleach, Wetwipes
3         Socks, Screws, Nails
4    Gymshorts, Hammer, Hinges

In [ ]:
#Since it's missing a header we'll create one, the new column header 'Products' 
df=pd.read_csv('lab5.csv', names=['products'])
df.head() 


products
0   Sanitizer, Wetwipes, Bread
1   Sanitizer, Wetwipes, Juice
2    Slippers, Wetwipes, Juice
3  Sanitizer, Bleach, Wetwipes
4         Socks, Screws, Nails

In [ ]:
df.shape

(7, 1)

In [ ]:
#the following code converts a pandas dataframe into a list of lists.
data=list(df["products"].apply(lambda x:x.split(",")))
data

[['Sanitizer', ' Wetwipes', ' Bread'],
 ['Sanitizer', ' Wetwipes', ' Juice'],
 ['Slippers', ' Wetwipes', ' Juice'],
 ['Sanitizer', ' Bleach', ' Wetwipes'],
 ['Socks', ' Screws', ' Nails'],
 ['Gymshorts', ' Hammer', ' Hinges'],
 ['Sanitizer', ' Juice', ' Wetwipes']]

In [ ]:
#Let's transform the list with one-hot encoding (make sure you understand the one-hot encoding concept first)
#Each column represents an item and each row represent a record or a transaction. 
#Initially, if that particular item occurs then it's labeled TRUE, if not then FALSE.
#After that, TRUE is converted to 1 and FALSE to 0 because computer loves numeric values.

from mlxtend.preprocessing import TransactionEncoder
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)

#Turn False & True to 0 & 1 for easy readability
df=df.replace(False,0)
df=df.replace(True,1)

df
                

Bleach   Bread   Hammer   Hinges   Juice   Nails   Screws   Wetwipes  \
0        0       1        0        0       0       0        0          1   
1        0       0        0        0       1       0        0          1   
2        0       0        0        0       1       0        0          1   
3        1       0        0        0       0       0        0          1   
4        0       0        0        0       0       1        1          0   
5        0       0        1        1       0       0        0          0   
6        0       0        0        0       1       0        0          1   

   Gymshorts  Sanitizer  Slippers  Socks  
0          0          1         0      0  
1          0          1         0      0  
2          0          0         1      0  
3          0          1         0      0  
4          0          0         0      1  
5          1          0         0      0  
6          0          1         0      0

In [ ]:
#We'll try 0.2 for the SUPPORT threshold value (but you can set it to any value you want).
# The parameter (use_colnames=True) is used to convert the integer back to its respective name.

df=apriori(df,min_support=0.2,use_colnames=True)
df

support                        itemsets
0  0.428571                        ( Juice)
1  0.714286                     ( Wetwipes)
2  0.571429                     (Sanitizer)
3  0.428571             ( Wetwipes,  Juice)
4  0.285714             ( Juice, Sanitizer)
5  0.571429          ( Wetwipes, Sanitizer)
6  0.285714  ( Wetwipes,  Juice, Sanitizer)

With the SUPPORT values above, you can try and match your own manual calculation. You notice how some are grouped with two. ALL items have been tested alone and in groups, but only these ones with the SUPPORT threshold 0.2 or higher values (i.e. other groups didn't make 0.2 therefore not shown).

In [ ]:
#Let's view our interpretation values using the Association rule function.
df_ar=association_rules(df,metric="lift", min_threshold=1)
df_ar

antecedents             consequents  antecedent support  \
0              ( Wetwipes)                ( Juice)            0.714286   
1                 ( Juice)             ( Wetwipes)            0.428571   
2                 ( Juice)             (Sanitizer)            0.428571   
3              (Sanitizer)                ( Juice)            0.571429   
4              ( Wetwipes)             (Sanitizer)            0.714286   
5              (Sanitizer)             ( Wetwipes)            0.571429   
6      ( Wetwipes,  Juice)             (Sanitizer)            0.428571   
7   ( Wetwipes, Sanitizer)                ( Juice)            0.571429   
8      ( Juice, Sanitizer)             ( Wetwipes)            0.285714   
9              ( Wetwipes)     ( Juice, Sanitizer)            0.714286   
10                ( Juice)  ( Wetwipes, Sanitizer)            0.428571   
11             (Sanitizer)     ( Wetwipes,  Juice)            0.571429   

    consequent support   support  confidence      lift  leverage  conviction  
0             0.428571  0.428571    0.600000  1.400000  0.122449    1.428571  
1             0.714286  0.428571    1.000000  1.400000  0.122449         inf  
2             0.571429  0.285714    0.666667  1.166667  0.040816    1.285714  
3             0.428571  0.285714    0.500000  1.166667  0.040816    1.142857  
4             0.571429  0.571429    0.800000  1.400000  0.163265    2.142857  
5             0.714286  0.571429    1.000000  1.400000  0.163265         inf  
6             0.571429  0.285714    0.666667  1.166667  0.040816    1.285714  
7             0.428571  0.285714    0.500000  1.166667  0.040816    1.142857  
8             0.714286  0.285714    1.000000  1.400000  0.081633         inf  
9             0.285714  0.285714    0.400000  1.400000  0.081633    1.190476  
10            0.571429  0.285714    0.666667  1.166667  0.040816    1.285714  
11            0.428571  0.285714    0.500000  1.166667  0.040816    1.142857



```
# This is formatted as code
```

From above result, we say that the support is 0.2 calculated by dividing the number of transactions containing Wetwipes and Juice by the total number of transactions.  The confidence level is 0.60 shows that out of all the transactions that contain juice 0.60 contain wetwipes too.  The lift 1.4 tells us that wetwipes is 1.4 times more likely to be bought by the customer who also buys juice.

In [ ]:
#Let's view our interpretation values using the Association rule function.
df_ar=association_rules(df,metric="confidence", min_threshold=0.6)
df_ar


antecedents             consequents  antecedent support  \
0          ( Wetwipes)                ( Juice)            0.714286   
1             ( Juice)             ( Wetwipes)            0.428571   
2             ( Juice)             (Sanitizer)            0.428571   
3          ( Wetwipes)             (Sanitizer)            0.714286   
4          (Sanitizer)             ( Wetwipes)            0.571429   
5  ( Wetwipes,  Juice)             (Sanitizer)            0.428571   
6  ( Juice, Sanitizer)             ( Wetwipes)            0.285714   
7             ( Juice)  ( Wetwipes, Sanitizer)            0.428571   

   consequent support   support  confidence      lift  leverage  conviction  
0            0.428571  0.428571    0.600000  1.400000  0.122449    1.428571  
1            0.714286  0.428571    1.000000  1.400000  0.122449         inf  
2            0.571429  0.285714    0.666667  1.166667  0.040816    1.285714  
3            0.571429  0.571429    0.800000  1.400000  0.163265    2.142857  
4            0.714286  0.571429    1.000000  1.400000  0.163265         inf  
5            0.571429  0.285714    0.666667  1.166667  0.040816    1.285714  
6            0.714286  0.285714    1.000000  1.400000  0.081633         inf  
7            0.571429  0.285714    0.666667  1.166667  0.040816    1.285714

From above result, let's take a look at the first row items: wetwipes and juice. 
The two items have the support of 0.20, where wetwipes is 0.43 and bread is 0.43. 
confidence is 1, this implies that if a person bought wetwipes then there is 100% juice is also bought. There are also positive correlations as we can see from the positive values of lift, leverage, and conviction.